### ISSUE TITLE
batch_size_per_gpu

### ISSUE THREAD

**Issue by clelevo** [sunny2109/SAFMN#18]
您好！我在训练时设置batch_size_per_gpu=8训练时间为16h，batch_size_per_gpu=48训练时间为16d，两者CPU均为满负载，是因为CPU IO瓶颈了嘛？将数据集读取格式为lmdb也没有影响。请问是哪里出了问题？

**Comment by clelevo** [comment]
且GPU负载均很低，约为25%，且呈现阶段性0负载

**Comment by bo-henry** [comment]
batch_size_per_gpu=64和batch_size_per_gpu=48为什么训练需要的epoch一样啊？

**Comment by sunny2109** [comment]
> 您好！我在训练时设置batch_size_per_gpu=8训练时间为16h，batch_size_per_gpu=48训练时间为16d，两者CPU均为满负载，是因为CPU IO瓶颈了嘛？将数据集读取格式为lmdb也没有影响。请问是哪里出了问题？

不好意思，回复晚了。你试一下加大number_worker，应该会有所缓解。

**Comment by clelevo** [comment]
> > 您好！我在训练时设置batch_size_per_gpu=8训练时间为16h，batch_size_per_gpu=48训练时间为16d，两者CPU均为满负载，是因为CPU IO瓶颈了嘛？将数据集读取格式为lmdb也没有影响。请问是哪里出了问题？
> 
> 不好意思，回复晚了。你试一下加大number_worker，应该会有所缓解。

非常感谢您的建议，我试着将number_worker从8调整到10（调整到更大就报错了），训练时间并没有什么变化。
设备信息：
CPU:Intel12700
GPU：NVIDIA 4090
OS: MS windows10
并且也尝试使用lmdb格式，运行速度反而更慢了。

**Comment by clelevo** [comment]
> > 您好！我在训练时设置batch_size_per_gpu=8训练时间为16h，batch_size_per_gpu=48训练时间为16d，两者CPU均为满负载，是因为CPU IO瓶颈了嘛？将数据集读取格式为lmdb也没有影响。请问是哪里出了问题？
> 
> 不好意思，回复晚了。你试一下加大number_worker，应该会有所缓解。

而且还有一个很奇怪的现象，训练大模型的时间反而比训练小模型的时间短

**Comment by clelevo** [comment]
> > 您好！我在训练时设置batch_size_per_gpu=8训练时间为16h，batch_size_per_gpu=48训练时间为16d，两者CPU均为满负载，是因为CPU IO瓶颈了嘛？将数据集读取格式为lmdb也没有影响。请问是哪里出了问题？
> 
> 不好意思，回复晚了。你试一下加大number_worker，应该会有所缓解。
在您贴出来的log中，第29200iter 才进入epoch1，而我训练时第600iter就进入epoch1

**Comment by sunny2109** [comment]
> > > 您好！我在训练时设置batch_size_per_gpu=8训练时间为16h，batch_size_per_gpu=48训练时间为16d，两者CPU均为满负载，是因为CPU IO瓶颈了嘛？将数据集读取格式为lmdb也没有影响。请问是哪里出了问题？
> > 
> > 
> > 不好意思，回复晚了。你试一下加大number_worker，应该会有所缓解。
> > 在您贴出来的log中，第29200iter 才进入epoch1，而我训练时第600iter就进入epoch1

你好，你提前把大图切成patch了吗？如果切了的话，应该会是一样的。
我这边训练小模型大概20h，大模型2天半左右。4090按理说速度应该更快。

**Comment by clelevo** [comment]
> > > > 您好！我在训练时设置batch_size_per_gpu=8训练时间为16h，batch_size_per_gpu=48训练时间为16d，两者CPU均为满负载，是因为CPU IO瓶颈了嘛？将数据集读取格式为lmdb也没有影响。请问是哪里出了问题？
> > > 
> > > 
> > > 不好意思，回复晚了。你试一下加大number_worker，应该会有所缓解。
> > > 在您贴出来的log中，第29200iter 才进入epoch1，而我训练时第600iter就进入epoch1
> 
> 你好，你提前把大图切成patch了吗？如果切了的话，应该会是一样的。 我这边训练小模型大概20h，大模型2天半左右。4090按理说速度应该更快。
您好，您说的把大图切成patch是指gt_size: 256嘛
X4.yml：
datasets:
  train:
    name: DF2K
    type: PairedImageDataset
    dataroot_gt: D:\PycharmProject\SAFMN-main\datasets\DF2K\DF2K_train_HR_sub
    dataroot_lq: D:\PycharmProject\SAFMN-main\datasets\DF2K\DF2K_train_LR_bicubic\X4_sub
#    dataroot_gt: D:\PycharmProject\SAFMN-main\datasets\DF2K\DF2K_train_HR.lmdb
#    dataroot_lq: D:\PycharmProject\SAFMN-main\datasets\DF2K\DF2K_train_LR_bicubic\X4.lmdb
    filename_tmpl: '{}'
    io_backend:
      type: disk

    gt_size: 256
    use_hflip: true
    use_rot: true

    # data loader    use_shuffle: true
    num_worker_per_gpu: 8
    batch_size_per_gpu: 64
    dataset_enlarge_ratio: 10
    prefetch_mode: ~
#    pin_memory: true

  # we use multiple validation datasets. The SR benchmark datasets can be download from: https://cv.snu.ac.kr/research/EDSR/benchmark.tar
  val:
    name: DIV2K_val100
    type: PairedImageDataset
    dataroot_gt: D:\PycharmProject\SAFMN-main\datasets\DIV2K\DIV2K_Valid\HR
    dataroot_lq: D:\PycharmProject\SAFMN-main\datasets\DIV2K\DIV2K_Valid\LR_bicubic\X4
    filename_tmpl: '{}x4'
    io_backend:
      type: disk

**Comment by clelevo** [comment]
> > > > 您好！我在训练时设置batch_size_per_gpu=8训练时间为16h，batch_size_per_gpu=48训练时间为16d，两者CPU均为满负载，是因为CPU IO瓶颈了嘛？将数据集读取格式为lmdb也没有影响。请问是哪里出了问题？
> > > 
> > > 
> > > 不好意思，回复晚了。你试一下加大number_worker，应该会有所缓解。
> > > 在您贴出来的log中，第29200iter 才进入epoch1，而我训练时第600iter就进入epoch1
> 
> 你好，你提前把大图切成patch了吗？如果切了的话，应该会是一样的。 我这边训练小模型大概20h，大模型2天半左右。4090按理说速度应该更快。

您好，您说的把大图切成patch是指gt_size: 256嘛？
我的X4.yml配置如下：
name: MIXSAFMNx4
model_type: SRModel
scale: 4
num_gpu: 1  
manual_seed: 10


datasets:
  train:
    name: DF2K
    type: PairedImageDataset
    dataroot_gt: D:\PycharmProject\SAFMN-main\datasets\DF2K\DF2K_train_HR_sub
    dataroot_lq: D:\PycharmProject\SAFMN-main\datasets\DF2K\DF2K_train_LR_bicubic\X4_sub

    filename_tmpl: '{}'
    io_backend:
      type: disk

    gt_size: 256
    use_hflip: true
    use_rot: true


    num_worker_per_gpu: 8
    batch_size_per_gpu: 64
    dataset_enlarge_ratio: 10
    prefetch_mode: ~


  val:
    name: DIV2K_val100
    type: PairedImageDataset
    dataroot_gt: D:\PycharmProject\SAFMN-main\datasets\DIV2K\DIV2K_Valid\HR
    dataroot_lq: D:\PycharmProject\SAFMN-main\datasets\DIV2K\DIV2K_Valid\LR_bicubic\X4
    filename_tmpl: '{}x4'
    io_backend:
      type: disk


network_g:
  type: MIXSAFMN
  dim: 48 
  n_blocks: 8
  ffn_scale: 2.0
  upscale: 4


path:
  pretrain_network_g: ~
  strict_load_g: false
  resume_state:


train:
  ema_decay: 0.999
  optim_g:
    type: AdamW
    lr: !!float 1e-3
    weight_decay: 0
    betas: [0.9, 0.99]

  scheduler:
    type: CosineAnnealingRestartLR
    periods: [500000]
    restart_weights: [1]
    eta_min: !!float 1e-4

  total_iter: 500000
  warmup_iter: -1  


  pixel_opt:
    type: L1Loss
    loss_weight: 1.0
    reduction: mean

  fft_opt:
    type: FFTLoss
    loss_weight: 0.05
    reduction: mean


val:
  val_freq: !!float 5000
  save_img: False
  pbar: False

  metrics:
    psnr:
      type: calculate_psnr
      crop_border: 2
      test_y_channel: False
      better: higher  

logger:
  print_freq: 100
  save_checkpoint_freq: !!float 5000
  use_tb_logger: true
  wandb:
    project: ~
    resume_id: ~


dist_params:
  backend: nccl
  port: 29500

**Comment by sunny2109** [comment]
> > > > > 您好！我在训练时设置batch_size_per_gpu=8训练时间为16h，batch_size_per_gpu=48训练时间为16d，两者CPU均为满负载，是因为CPU IO瓶颈了嘛？将数据集读取格式为lmdb也没有影响。请问是哪里出了问题？
> > > > 
> > > > 
> > > > 不好意思，回复晚了。你试一下加大number_worker，应该会有所缓解。
> > > > 在您贴出来的log中，第29200iter 才进入epoch1，而我训练时第600iter就进入epoch1
> > 
> > 
> > 你好，你提前把大图切成patch了吗？如果切了的话，应该会是一样的。 我这边训练小模型大概20h，大模型2天半左右。4090按理说速度应该更快。
> 
> 您好，您说的把大图切成patch是指gt_size: 256嘛？ 我的X4.yml配置如下： name: MIXSAFMNx4 model_type: SRModel scale: 4 num_gpu: 1 manual_seed: 10
> 
> datasets: train: name: DF2K type: PairedImageDataset dataroot_gt: D:\PycharmProject\SAFMN-main\datasets\DF2K\DF2K_train_HR_sub dataroot_lq: D:\PycharmProject\SAFMN-main\datasets\DF2K\DF2K_train_LR_bicubic\X4_sub
> 
> ```
> filename_tmpl: '{}'
> io_backend:
>   type: disk
> 
> gt_size: 256
> use_hflip: true
> use_rot: true
> 
> 
> num_worker_per_gpu: 8
> batch_size_per_gpu: 64
> dataset_enlarge_ratio: 10
> prefetch_mode: ~
> ```
> 
> val: name: DIV2K_val100 type: PairedImageDataset dataroot_gt: D:\PycharmProject\SAFMN-main\datasets\DIV2K\DIV2K_Valid\HR dataroot_lq: D:\PycharmProject\SAFMN-main\datasets\DIV2K\DIV2K_Valid\LR_bicubic\X4 filename_tmpl: '{}x4' io_backend: type: disk
> 
> network_g: type: MIXSAFMN dim: 48 n_blocks: 8 ffn_scale: 2.0 upscale: 4
> 
> path: pretrain_network_g: ~ strict_load_g: false resume_state:
> 
> train: ema_decay: 0.999 optim_g: type: AdamW lr: !!float 1e-3 weight_decay: 0 betas: [0.9, 0.99]
> 
> scheduler: type: CosineAnnealingRestartLR periods: [500000] restart_weights: [1] eta_min: !!float 1e-4
> 
> total_iter: 500000 warmup_iter: -1
> 
> pixel_opt: type: L1Loss loss_weight: 1.0 reduction: mean
> 
> fft_opt: type: FFTLoss loss_weight: 0.05 reduction: mean
> 
> val: val_freq: !!float 5000 save_img: False pbar: False
> 
> metrics: psnr: type: calculate_psnr crop_border: 2 test_y_channel: False better: higher
> 
> logger: print_freq: 100 save_checkpoint_freq: !!float 5000 use_tb_logger: true wandb: project: ~ resume_id: ~
> 
> dist_params: backend: nccl port: 29500

就是先把[2K的图像裁成480*480的patch](https://github.com/sunny2109/SAFMN/blob/main/scripts/data_preparation/extract_subimages.py)，这样读取就不用读整图，速度快很多。

**Comment by clelevo** [comment]
> > > > > > 您好！我在训练时设置batch_size_per_gpu=8训练时间为16h，batch_size_per_gpu=48训练时间为16d，两者CPU均为满负载，是因为CPU IO瓶颈了嘛？将数据集读取格式为lmdb也没有影响。请问是哪里出了问题？
> > > > > 
> > > > > 
> > > > > 不好意思，回复晚了。你试一下加大number_worker，应该会有所缓解。
> > > > > 在您贴出来的log中，第29200iter 才进入epoch1，而我训练时第600iter就进入epoch1
> > > 
> > > 
> > > 你好，你提前把大图切成patch了吗？如果切了的话，应该会是一样的。 我这边训练小模型大概20h，大模型2天半左右。4090按理说速度应该更快。
> > 
> > 
> > 您好，您说的把大图切成patch是指gt_size: 256嘛？ 我的X4.yml配置如下： name: MIXSAFMNx4 model_type: SRModel scale: 4 num_gpu: 1 manual_seed: 10
> > datasets: train: name: DF2K type: PairedImageDataset dataroot_gt: D:\PycharmProject\SAFMN-main\datasets\DF2K\DF2K_train_HR_sub dataroot_lq: D:\PycharmProject\SAFMN-main\datasets\DF2K\DF2K_train_LR_bicubic\X4_sub
> > ```
> > filename_tmpl: '{}'
> > io_backend:
> >   type: disk
> > 
> > gt_size: 256
> > use_hflip: true
> > use_rot: true
> > 
> > 
> > num_worker_per_gpu: 8
> > batch_size_per_gpu: 64
> > dataset_enlarge_ratio: 10
> > prefetch_mode: ~
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > val: name: DIV2K_val100 type: PairedImageDataset dataroot_gt: D:\PycharmProject\SAFMN-main\datasets\DIV2K\DIV2K_Valid\HR dataroot_lq: D:\PycharmProject\SAFMN-main\datasets\DIV2K\DIV2K_Valid\LR_bicubic\X4 filename_tmpl: '{}x4' io_backend: type: disk
> > network_g: type: MIXSAFMN dim: 48 n_blocks: 8 ffn_scale: 2.0 upscale: 4
> > path: pretrain_network_g: ~ strict_load_g: false resume_state:
> > train: ema_decay: 0.999 optim_g: type: AdamW lr: !!float 1e-3 weight_decay: 0 betas: [0.9, 0.99]
> > scheduler: type: CosineAnnealingRestartLR periods: [500000] restart_weights: [1] eta_min: !!float 1e-4
> > total_iter: 500000 warmup_iter: -1
> > pixel_opt: type: L1Loss loss_weight: 1.0 reduction: mean
> > fft_opt: type: FFTLoss loss_weight: 0.05 reduction: mean
> > val: val_freq: !!float 5000 save_img: False pbar: False
> > metrics: psnr: type: calculate_psnr crop_border: 2 test_y_channel: False better: higher
> > logger: print_freq: 100 save_checkpoint_freq: !!float 5000 use_tb_logger: true wandb: project: ~ resume_id: ~
> > dist_params: backend: nccl port: 29500
> 
> 就是先把[2K的图像裁成480*480的patch](https://github.com/sunny2109/SAFMN/blob/main/scripts/data_preparation/extract_subimages.py)，这样读取就不用读整图，速度快很多。我不确定你HR_sub是不是用的patch，如果是的话，i/o应该跟得上才对。

太感谢您了，我并没有使用extract_subimages.py将2K图像裁剪成480*480。
祝您生活愉快~~~
